ADVANCED_EXAMPLES.md

```diff
@@ -0,0 +1,596 @@
+- [Mocking in the unit tests](#mocking-in-the-unit-tests)
+- [Hacking stamps](#hacking-stamps)
+  - [Enforced default properties](#enforced-default-properties)
+  - [Enforced default properties as a composable behavior](#enforced-default-properties-as-a-composable-behavior)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+
+## Self aware objects - `instance.getStamp()`. 2 ways.
+
+> Run the examples below for yourself:
+```sh
+$ git clone https://github.com/stampit-org/stampit.git
+$ cd stampit && npm i babel -g
+$ babel-node advanced-examples/self-aware.js
+```
+
+You can add `.getStamp()` function to each object with ease. 
+
+First, let's assume you have a following stamp:
+```js
+const User = stampit();
+```
+
+### Attach function to object
+Just compose the following stamp to any other stamp.
+```js
+const SelfAware1 = stampit.init(({ instance, stamp }) => {
+  instance.getStamp = () => stamp;
+});
+```
+Let's compose it with the `User` stamp from above:
+```js
+const SelfAwareUser1 = User.compose(SelfAware1);
+```
+Now, let's create a user and call the `.getStamp()`:
+```js
+const user1 = SelfAwareUser1();
+assert.strictEqual(user1.getStamp(), SelfAwareUser1); // All good
+```
+So, now every object instance returns the exact stamp it was built with. Nice!
+
+### Attach function to prototype (memeory efficient)
+Another composable stamp which does the same but in a memory efficient way.
+It attaches the function to the `.prototype` of the objects, but not to each one.
+```js
+const SelfAware2 = stampit.init(({ instance, stamp }) => {
+  if (instance.getStamp) {
+    stamp.fixed.methods.getStamp = () => stamp;
+  }
+});
+```
+The `ctx.stamp.fixed` property contains stamp's internal data.
+The `fixed.methods` object is used as all object instances' `.prototype`.
+
+Compose this new stamp with our `User` from above:
+```js
+const SelfAwareUser2 = User.compose(SelfAware2);
+```
+Let's test it:
+```js
+const user2 = SelfAwareUser2();
+assert.strictEqual(user2.getStamp(), SelfAwareUser2); // All good
+```
+And again, every new object instance knows which stamp it was made of. Brilliant!
+
+------------------------------
+
+## Self cloneable objects - `instance.clone()`. 3 ways.
+
+> Run the examples below for yourself:
+```sh
+$ git clone https://github.com/stampit-org/stampit.git
+$ cd stampit && npm i babel -g
+$ babel-node advanced-examples/cloneable.js
+```
+
+This is simple stamp with a single method and a single state property `prefix`.
+```js
+const PrependLogger = stampit.methods({
+  log: function (obj) {
+    console.log(this.prefix, obj);
+  }
+}).refs({
+  prefix: 'STDOUT: '
+});
+```
+Using it:
+```js
+const originalLogger = PrependLogger();
+originalLogger.log('hello');
+```
+Prints `STDOUT: hello`
+
+### Attach method to each object instance
+
+Let's implement a stamp which allows **any** object to be safely cloned: 
+```js
+const Cloneable1 = stampit.init(({ instance, stamp }) => {
```

Comment:
No need for the extra parenthesis

`({ instance, stamp })` -> `{ instance, stamp }`